### PR TITLE
Feat/ Add wallet - Core implementation

### PR DESCRIPTION
**Pull request by fredericvilcot**
This PR carries the first stage of construction of the `Wallet` organism by building in particular the architectural design.

***Architecture***

Wallet is built upon [hexagonal architecture](https://blog.octo.com/architecture-hexagonale-trois-principes-et-un-exemple-dimplementation/) that allows to decouple all business logic from UI framework and ensures maximum testability and scalability.
In this feature, this kind of architecture allows to switch from any kind of wallet and to systematically provide the user with the same associated service.

***Purpose***

This first implementation allows user to enable [Keplr wallet](https://wallet.keplr.app/#/dashboard) and to retrieve all associated accounts that are represented by address / pubKey pairs per `chainId`

### CODE REVIEW

**Review by ad2ien — APPROVED** [review]
Very cool!

**Review by amimart — DISMISSED** [review]

**Review by ccamel — CHANGES_REQUESTED** [review]
Good job! 👍 

A few small remarks, but nothing major, which I would be happy to discuss with you. Thanks.

File: src/gateways/KeplrWalletGateway.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+export class KeplrWallet implements Wallet {
+  public isConnected = (): boolean =>
+    !!window.Keplr.getOfflineSigner && !!window.Keplr
+
+  public connect = (chainId: ChainId): Promise<void | ConnectionError> => {
+    if (!this.isConnected) {
+      return Promise.resolve(new ConnectionError())
+    }
+    return window.Keplr.enable(chainId)
```
Comment (ccamel) [inline src/gateways/KeplrWalletGateway.ts:15]: Tested in the console, the `enable()` can throw exceptions (I don't think it's clearly described in the official documentatin). So better wrap with a try/catch construct.

```
> window.keplr.enable("foo")
Promise {<pending>}
injectedScript.bundle.js:10 Uncaught (in promise) Error: There is no chain info for foo
    at r (injectedScript.bundle.js:10:32583)
```

File: src/gateways/DTO/keplrDTO.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { AccountData } from 'cosmjs/launchpad'
+
+export type KeplrAccountDTO = AccountData
```
[the remark this answers is not in the document]
Comment (fredericvilcot) [reply src/gateways/DTO/keplrDTO.ts:3]: It was to enhance the mapping operation but I agree, it's unnecessary here ...

File: src/domain/wallet/builders/account.builder.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  constructor() {
+    this.account = {
+      address: '',
+      publicKey: new Uint8Array(),
+    }
+  }
+
+  public withAddress(address: string): AccountBuilder {
+    this.account.address = address
```
[the remark this answers is not in the document]
Comment (fredericvilcot) [reply src/domain/wallet/builders/account.builder.ts:14]: Thanks for highlighting this elements, it makes me understand better the final purpose of `builder` pattern !

File: src/gateways/DTO/keplrDTO.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccamel) [reply src/gateways/DTO/keplrDTO.ts:3]: Thanks 😌

File: src/domain/wallet/builders/account.builder.spec.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    const pubKey = new Uint8Array(1)
+    describe('When building account', () => {
+      const account = () => buildAccount(address, pubKey)
+      test('Then, expect builder to throw an UnspecifiedError ', () => {
+        expect(account).toThrowError(UnspecifiedError)
+      })
+    })
+  })
+
+  describe('Given that publicKey is empty', () => {
```
[the remark this answers is not in the document]
Comment (ccamel) [reply src/domain/wallet/builders/account.builder.spec.ts:19]: As discussed together @fredericvilcot not relevant to me. Let's find a way to disable that rule.

File: src/domain/wallet/builders/account.builder.spec.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import { UnspecifiedError } from '../entities/errors'
+import { AccountBuilder } from './account.builder'
+
+describe('Build an Account', () => {
+  const buildAccount = (adress: string, pubKey: Uint8Array) =>
+    new AccountBuilder().withAddress(adress).withPublicKey(pubKey).build()
+
+  describe('Given that address is empty', () => {
```
[the remark this answers is not in the document]
Comment (ccamel) [reply src/domain/wallet/builders/account.builder.spec.ts:8]: As discussed together @fredericvilcot not relevant to me. Let's find a way to disable that rule.

**Review by ccamel — APPROVED** [review]
Amazing work! 😻

**Review by amimart — APPROVED** [review]

**Pull request merged**